File: src/ekf_core.py
"""
ekf_core.py
===========
15-state Extended Kalman Filter for UAV INS.
Enhanced version with IMU bias estimation.

State vector (15×1):
  x = [px, py, pz,          <- Position (NED, m)
       vx, vy, vz,          <- Velocity (NED, m/s)
       phi, theta, psi,     <- Euler Angles (rad)
       abx, aby, abz,       <- Accel Bias (m/s²)
       gbx, gby, gbz]       <- Gyro Bias (rad/s)
"""

import numpy as np
import math
from imu_noise_params import IMUNoiseParams

GRAVITY = np.array([0.0, 0.0, 9.80665])

class EKFCore:
    def __init__(self, noise: IMUNoiseParams):
        self.noise = noise
        
        # 1. Initialize State Vector (15 elements)
        self.x = np.zeros(15)
        
        # 2. Initialize Covariance Matrix (15x15)
        # We start with small uncertainties for position and velocity,
        # and very small ones for biases.
        self.P = np.eye(15)
        self.P[0:3, 0:3] *= 1.0     # Position (m²)
        self.P[3:6, 3:6] *= 0.1     # Velocity (m/s)²
        self.P[6:9, 6:9] *= 0.01    # Attitude (rad²)
        self.P[9:12, 9:12] *= 0.01  # Accel bias (m/s²)²
        self.P[12:15, 12:15] *= 0.001 # Gyro bias (rad/s)²

        # 3. Process Noise Q (15x15)
        # Models how much we "mistrust" our prediction over time.
        self.Q = np.zeros((15, 15))
        sa = noise.accel_std ** 2
        sg = noise.gyro_std  ** 2
        sab = 1e-6 # Accel bias stability
        sgb = 1e-8 # Gyro bias stability
        
        # Diagonal process noise elements
        np.fill_diagonal(self.Q[3:6, 3:6], sa)     # Velocity driven by accel noise
        np.fill_diagonal(self.Q[6:9, 6:9], sg)     # Attitude driven by gyro noise
        np.fill_diagonal(self.Q[9:12, 9:12], sab)  # Accel bias drift
        np.fill_diagonal(self.Q[12:15, 12:15], sgb)# Gyro bias drift

        # 4. Measurement Noise
        self.R_baro = np.array([[noise.baro_std ** 2]])
        self.R_mag  = np.array([[noise.mag_std ** 2]])

        # 5. Observation Matrices
        self.H_baro = np.zeros((1, 15))
        self.H_baro[0, 2] = 1.0  # Baro observes pz
        
        self.H_mag = np.zeros((1, 15))
        self.H_mag[0, 8] = 1.0   # Mag observes psi

    @property
    def state(self) -> dict:
        return {
            "pos":   self.x[0:3].copy(),
            "vel":   self.x[3:6].copy(),
            "euler": self.x[6:9].copy(),
            "accel_bias": self.x[9:12].copy(),
            "gyro_bias":  self.x[12:15].copy()
        }

    def predict(self, accel_raw: np.ndarray, gyro_raw: np.ndarray, dt: float):
        """
        Predict step: Propagates the state forward using IMU mechanization.
        Now compensates for estimated biases.
        """
        # --- A. Compensate Biases ---
        accel = accel_raw - self.x[9:12]
        gyro  = gyro_raw  - self.x[12:15]
        
        phi, theta, psi = self.x[6:9]
        R = self._R_bn(phi, theta, psi)
        T = self._T_inv(phi, theta)

        # --- B. State Propagation ---
        # 1. Velocity Update (v_new = v + (R*f + g)*dt)
        f_ned = R @ accel
        acc_ned = f_ned + GRAVITY
        self.x[3:6] += acc_ned * dt

        # 2. Position Update (p_new = p + v*dt)
        self.x[0:3] += self.x[3:6] * dt

        # 3. Attitude Update (eta_new = eta + (T*omega)*dt)
        eta_dot = T @ gyro
        self.x[6:9] += eta_dot * dt
        self.x[6:9] = [self._wrap_angle(a) for a in self.x[6:9]]

        # --- C. Covariance Propagation (P = FPF' + Q) ---
        F = self._compute_F_15(accel, phi, theta, psi, dt)
        self.P = F @ self.P @ F.T + self.Q * dt

    def _compute_F_15(self, accel, phi, theta, psi, dt) -> np.ndarray:
        """Linearized state transition Jacobian (15x15)"""
        F = np.eye(15)
        
        # Pos derivative w.r.t Vel
        F[0:3, 3:6] += np.eye(3) * dt
        
        # Vel derivative w.r.t Euler (Numerical)
        R = self._R_bn(phi, theta, psi)
        f_ned = R @ accel
        eps = 1e-5
        for i in range(3):
            euler_p = np.array([phi, theta, psi])
            euler_p[i] += eps
            df = (self._R_bn(*euler_p) @ accel - f_ned) / eps
            F[3:6, 6+i] = df * dt
            
        # Vel derivative w.r.t Accel Bias
        F[3:6, 9:12] = -R * dt
        
        # Euler derivative w.r.t Gyro Bias
        T = self._T_inv(phi, theta)
        F[6:9, 12:15] = -T * dt
        
        return F

    def update_baro(self, alt_measured: float):
        self.x[6:9] = [self._wrap_angle(a) for a in self.x[6:9]]

    # ═══════════════════════════════════════════════════════════
    # UPDATE — magnetometer (yaw)
    # ═══════════════════════════════════════════════════════════
    def update_mag(self, yaw_measured: float):
        """
        EKF measurement update using magnetometer yaw.
        H = [0 0 0  0 0 0  0 0 1]  (observes psi)
        Innovation is wrapped to [-π, π] to avoid wrap-around jumps.
        """
        z_meas = np.array([yaw_measured])
        z_pred = np.array([self.x[8]])

        y = np.array([self._wrap_angle(z_meas[0] - z_pred[0])])

        S = self.H_mag @ self.P @ self.H_mag.T + self.R_mag
        K = self.P @ self.H_mag.T @ np.linalg.inv(S)

        self.x = self.x + (K @ y).flatten()
        self.P = (np.eye(9) - K @ self.H_mag) @ self.P

        self.x[6:9] = [self._wrap_angle(a) for a in self.x[6:9]]

    # ── util ───────────────────────────────────────────────────
    @staticmethod
    def _wrap_angle(a: float) -> float:
        """Wrap angle to [-π, π]."""
        import math
        return math.atan2(math.sin(a), math.cos(a))

    def reset(self):
        """Reset state and covariance to initial values."""
        self.__init__(self.noise)

File: INS SYSTEM SIMULATED USING THE MATLAB/simulation_source/ekf_core.m
%% =========================================================
%  Extended Kalman Filter — 15-State Core Functions
%  =========================================================
%  State vector x (15×1):
%    x(1:3)   = position        [m] (NED)
%    x(4:6)   = velocity        [m/s] (NED)
%    x(7:9)   = Euler angles    [rad] (roll, pitch, yaw)
%    x(10:12) = Accel bias      [m/s²]
%    x(13:15) = Gyro bias       [rad/s]
% =========================================================

function [x, P] = ekf_init(pos0, vel0, euler0)
    % Initialise EKF with 15 states
    x = zeros(15, 1);
    x(1:3) = pos0;
    x(4:6) = vel0;
    x(7:9) = euler0;
    
    % Initial covariance: trust position/velocity/euler, assume small bias
    P = diag([1,1,1, 0.5,0.5,0.5, 0.1,0.1,0.1, 0.01,0.01,0.01, 0.001,0.001,0.001].^2);
end

function [x_new, P_new] = ekf_predict(x, P, accel_raw, gyro_raw, dt, imu)
    % 1. Extract states
    pos   = x(1:3);
    vel   = x(4:6);
    euler = x(7:9);
    ba    = x(10:12); % Accel bias
    bg    = x(13:15); % Gyro bias
    
    % 2. Compensate for Biases
    accel = accel_raw - ba;
    gyro  = gyro_raw - bg;
    
    roll = euler(1); pitch = euler(2); yaw = euler(3);
    R_bn = rot_body2ned(roll, pitch, yaw);
    g_ned = [0; 0; 9.81];
    
    % 3. State Propagation (Kinematics)
    a_ned = R_bn * accel + g_ned;
    vel_new = vel + a_ned * dt;
    pos_new = pos + vel * dt + 0.5 * a_ned * dt^2;
    
    % Euler kinematics
    cr = cos(roll); sr = sin(roll); cp = cos(pitch); sp = sin(pitch);
    T_inv = [1, sr*sp/cp, cr*sp/cp;
             0, cr,       -sr;
             0, sr/cp,    cr/cp];
    euler_dot = T_inv * gyro;
    euler_new = angle_wrap(euler + euler_dot * dt);
    
    x_new = [pos_new; vel_new; euler_new; ba; bg]; % Biases assumed constant (random walk in Q)
    
    % 4. Jacobian F (15x15)
    F = eye(15);
    F(1:3, 4:6) = eye(3) * dt;
    
    % Velocity w.r.t Euler
    da_dr = R_bn * skew([1;0;0]) * accel * dt;
    da_dp = R_bn * skew([0;1;0]) * accel * dt;
    da_dy = R_bn * skew([0;0;1]) * accel * dt;
    F(4:6, 7) = da_dr;
    F(4:6, 8) = da_dp;
    F(4:6, 9) = da_dy;
    
    % Velocity w.r.t Accel Bias
    F(4:6, 10:12) = -R_bn * dt;
    
    % Euler w.r.t Gyro Bias
    F(7:9, 13:15) = -T_inv * dt;
    
    % 5. Covariance Propagation
    % imu.Q should be 15x15 (or padded)
    Q15 = eye(15) * 1e-6; 
    Q15(4:6, 4:6) = eye(3) * (imu.accel_std^2);
    Q15(7:9, 7:9) = eye(3) * (imu.gyro_std^2);
    
    P_new = F * P * F' + Q15 * dt;
end

function [x_new, P_new] = ekf_update_baro(x, P, z_baro, imu)
    H = zeros(1, 15);
    H(3) = 1; % Observe NED z-position
    
    S = H * P * H' + imu.R_baro;
    K = (P * H') / S;
    
    innov = z_baro - x(3);
    x_new = x + K * innov;
    x_new(7:9) = angle_wrap(x_new(7:9));
    P_new = (eye(15) - K*H) * P;
end

function [x_new, P_new] = ekf_update_mag(x, P, z_yaw, imu)
    H = zeros(1, 15);
    H(9) = 1; % Observe yaw
    
    S = H * P * H' + imu.R_mag;
    K = (P * H') / S;
    
    innov = angle_wrap(z_yaw - x(9));
    x_new = x + K * innov;
    x_new(7:9) = angle_wrap(x_new(7:9));
    P_new = (eye(15) - K*H) * P;
end

% ── Helpers ──────────────────────────────────────────────
function R = rot_body2ned(r,p,y)
    cr=cos(r);sr=sin(r);cp=cos(p);sp=sin(p);cy=cos(y);sy=sin(y);
    R=[cp*cy-sr*sp*sy, -cr*sy, sp*cy+sr*cp*sy;
       cp*sy+sr*sp*cy,  cr*cy, sp*sy-sr*cp*cy;
       -cr*sp,          sr,    cr*cp];
end

function S = skew(v)
    S=[0,-v(3),v(2); v(3),0,-v(1); -v(2),v(1),0];
end

function a = angle_wrap(a)
    a = mod(a+pi, 2*pi) - pi;
end
